src/transformer.ts

```diff
@@ -534,3 +605,280 @@ export default {
     sheets,
     escapeURLPath,
 };
+
+const correctHTMLDocHierarchy = (doc: DOM.Document) => {
+    const { documentElement } = doc;
+
+    if (documentElement.nodeName.toLowerCase() === 'html') {
+        const root = getNodeByXPathExpression(doc, '/html/body/root');
+
+        if (root == null) {
+            throw new Error('Missing root node.');
+        }
+
+        documentElement.replaceWith(root);
+    }
+};
+
+/** @see {@link processItemsets} */
+const substringBefore = (haystack: string, needle: string) =>
+    haystack.split(needle, 1)[0];
+
+/** @see {@link processItemsets} */
+const substringAfter = (haystack: string, needle: string) =>
+    haystack.substring(haystack.indexOf(needle) + needle.length);
+
+/**
+ * This is a replacement for the XSL template named `strip-filter`. The original
+ * XSL commentary follows:
+ *
+ * ----------------------------------------------------------------------------
+ * turns: /path/to/node[value=/some/other/node] into: /path/to/node this
+ * function is probably way too aggressive but will work for xls-form generated
+ * forms to do this properly a regexp:replace is required, but not supported in
+ * libXML kept the recursion in, even though it is not being used right now
+ * ----------------------------------------------------------------------------
+ *
+ * Contrary to the JSDoc comment for @see {@link processItemsets},
+ * this implementation deviates from the original XSL logic, which uses
+ * substrings and recursion, because it's (hopefully) much easier to understand
+ * this way.
+ *
+ * It is *also* worth noting that in the process of migrating to browser DOM,
+ * the original commentary was found to be wrong! The recursion was certainly in
+ * use, and without it some snapshot tests failed.
+ */
+const stripFilter = (expression: string) => expression.replace(/\[.*?\]/g, '');
+
+/**
+ * This is a replacement for the dynamic logic in the XSL template
+ * `match="xf:itemset" mode="labels"`. Moving it to DOM code allows us to drop
+ * our reliance on the unsupported `dyn` extension. The intent is to match the
+ * naming and semantics of the original XSL logic to ensure there is no
+ * potential for regressions.
+ */
+const processItemsets = (xformDoc: DOM.Document) => {
```

Comment:
Is there a test of a select on a repeat? I can't quickly find one and I think it would be good to make sure there is.

Reply:
@eyelidlessness walked me through this more and I see now that it only applies to label expressions with `itext`.